provider/pkg/resources/cfn_custom_resource_test.go

```diff
@@ -0,0 +1,1540 @@
+				"stackId":      resource.NewStringProperty("testProject"),
+				"customResourceProperties": resource.NewObjectProperty(resource.NewPropertyMapFromMap(map[string]interface{}{
+					"level1": map[string]interface{}{
+						"level2": []interface{}{
+							map[string]interface{}{
+								"key1": "value1",
+								"key2": 2,
+							},
+							3.14,
+							"string",
+						},
+						"anotherKey": true,
+						"arrayOfMaps": []interface{}{
+							map[string]interface{}{
+								"key1": "value1",
+								"key2": 2,
+							},
+							map[string]interface{}{
+								"key3": "value3",
+								"key4": 4,
+							},
+						},
+					},
+				})),
+			},
+			expectedInputs: resource.PropertyMap{
+				"serviceToken": resource.NewStringProperty("arn:aws:lambda:us-west-2:123456789012:function:my-function"),
+				"stackId":      resource.NewStringProperty("testProject"),
+				"customResourceProperties": resource.NewObjectProperty(resource.NewPropertyMapFromMap(map[string]interface{}{
+					"level1": map[string]interface{}{
+						"level2": []interface{}{
+							map[string]interface{}{
+								"key1": "value1",
+								"key2": "2",
+							},
+							"3.14",
+							"string",
+						},
+						"anotherKey": "true",
+						"arrayOfMaps": []interface{}{
+							map[string]interface{}{
+								"key1": "value1",
+								"key2": "2",
+							},
+							map[string]interface{}{
+								"key3": "value3",
+								"key4": "4",
+							},
+						},
+					},
+				})),
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			c := &cfnCustomResource{}
+			ctx := context.Background()
+			urn := urn.URN("urn:pulumi:testProject::test::aws-native:cloudformation:CfnCustomResource::dummy")
+			randomSeed := []byte{}
+			state := resource.PropertyMap{}
+			defaultTags := map[string]string{}
+
+			newInputs, failures, err := c.Check(ctx, urn, randomSeed, tt.inputs, state, defaultTags)
+
+			if tt.expectedError != nil {
+				assert.Error(t, err)
+				assert.Contains(t, err.Error(), *tt.expectedError)
+			} else {
+				require.NoError(t, err)
+			}
+
+			assert.Equal(t, tt.expectedFailures, failures)
+
+			if tt.expectedInputs != nil {
+				assert.Equal(t, tt.expectedInputs, newInputs)
+			}
+		})
+	}
+}
+
+func TestCfnCustomResource_Create(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name                 string
+		timeout              time.Duration
+		noEcho               bool
+		expectedError        string
+		customResourceData   map[string]interface{}
+		customResourceInputs map[string]interface{}
+	}{
+		{
+			name: "Success",
+			customResourceData: map[string]interface{}{
+				"prop1": "value1",
+				"prop2": true,
+				"prop3": []interface{}{"a", "b", "c"},
+				"prop4": map[string]interface{}{
+					"nestedProp1": "nestedValue1",
+					"nestedProp2": 42,
+				},
+			},
+			customResourceInputs: map[string]interface{}{
+				"key": "value",
+			},
+		},
+		{
+			name:   "SecretResponse",
```

Comment:
Would it be possible to see here what is returned to the engine from Create in this test case?

Perhaps with `hexops/autogold` if it's too much trouble to write by hand.

Reply:
We're asserting this with the assertion in line 348. In case `noEcho` is set to true, we expect the `data` prop to be marked as a secret.

Reply:
Ah ok!